packages/devtools_app/lib/src/charts/flutter/flame_chart.dart

```diff
@@ -60,50 +63,98 @@ abstract class FlameChartState<T extends FlameChart, V> extends State<T>
 
-  LinkedScrollControllerGroup _linkedScrollControllerGroup;
-
   final List<FlameChartRow> rows = [];
 
   final List<FlameChartSection> sections = [];
 
+  ScrollController verticalScrollController;
+
+  LinkedScrollControllerGroup linkedHorizontalScrollControllerGroup;
+
+  double verticalScrollOffset = 0.0;
+
+  double horizontalScrollOffset = 0.0;
+
+  double zoom = minZoomLevel;
+
   /// Starting pixels per microsecond in order to fit all the data in view at
   /// start.
   double get startingPxPerMicro =>
       widget.startingContentWidth / widget.data.time.duration.inMicroseconds;
 
   int get startTimeOffset => widget.data.time.start.inMicroseconds;
 
+  /// Provides CustomPaint widgets to be painted on top of the flame chart, if
+  /// overridden.
+  ///
+  /// The painters will be painted in the order that they are returned.
+  List<CustomPaint> buildCustomPaints(BoxConstraints constraints) => [];
+
   @override
   void initState() {
     super.initState();
     initFlameChartElements();
-    _linkedScrollControllerGroup = LinkedScrollControllerGroup();
+
+    linkedHorizontalScrollControllerGroup = LinkedScrollControllerGroup();
+    linkedHorizontalScrollControllerGroup.onOffsetChanged(() {
+      setState(() {
+        horizontalScrollOffset = linkedHorizontalScrollControllerGroup.offset;
+      });
+    });
+    verticalScrollController = ScrollController();
+    verticalScrollController.addListener(() {
+      if (verticalScrollOffset != verticalScrollController.offset) {
+        setState(() {
+          verticalScrollOffset = verticalScrollController.offset;
+        });
+      }
+    });
   }
 
   @override
   void didUpdateWidget(T oldWidget) {
     if (widget.data != oldWidget.data) {
       initFlameChartElements();
-      _linkedScrollControllerGroup.resetScroll();
+      linkedHorizontalScrollControllerGroup.resetScroll();
+      verticalScrollController.jumpTo(0.0);
     }
     super.didUpdateWidget(oldWidget);
   }
 
+  @override
+  void dispose() {
+    verticalScrollController.dispose();
+    super.dispose();
+  }
+
   @override
   Widget build(BuildContext context) {
     return LayoutBuilder(
       builder: (context, constraints) {
-        return ListView.builder(
-          addAutomaticKeepAlives: false,
-          itemCount: rows.length,
-          itemBuilder: (context, index) {
-            return ScrollingFlameChartRow<V>(
-              linkedScrollControllerGroup: _linkedScrollControllerGroup,
-              nodes: rows[index].nodes,
-              width: math.max(constraints.maxWidth, widget.totalStartingWidth),
-              selected: widget.selected,
-            );
-          },
+        final customPaints = buildCustomPaints(constraints);
+        return customPaints.isNotEmpty
+            ? Stack(
+                children: [
+                  _buildFlameChart(constraints),
+                  ...customPaints,
+                ],
+              )
+            : _buildFlameChart(constraints);
```

Comment:
nit: also define a local variable for
_buildFlameChart(constraints)
so that isn't duplicated.

Reply:
done